hass_nabucasa/files.py

```diff
@@ -0,0 +1,210 @@
+                method="GET",
+                path="/upload_details",
+                jsondata={
+                    "storage_type": storage_type,
+                    "filename": filename,
+                    "md5": base64md5hash,
+                    "size": size,
+                    "metadata": metadata,
+                },
+            )
+        except TimeoutError as err:
+            raise FilesError(
+                "Timeout reached while trying to fetch upload details",
+            ) from err
+        except ClientError as err:
+            raise FilesError(f"Failed to fetch upload details: {err}") from err
+        except Exception as err:
+            raise FilesError(
+                f"Unexpected error while fetching upload details: {err}",
+            ) from err
+
+        try:
+            response = await self._cloud.websession.put(
+                details["url"],
+                data=await open_stream(),
+                headers=details["headers"] | {"content-length": str(size)},
+                timeout=ClientTimeout(connect=10.0, total=_FILE_TRANSFER_TIMEOUT),
+            )
+            self._do_log_response(response)
+
+            response.raise_for_status()
+        except TimeoutError as err:
+            raise FilesError("Timeout reached while trying to upload file") from err
+        except ClientError as err:
+            raise FilesError("Failed to upload file") from err
+        except Exception as err:
+            raise FilesError("Unexpected error while uploading file") from err
+
+    async def download(
+        self,
+        storage_type: StorageType,
+        filename: str,
+    ) -> StreamReader:
+        """Download a file."""
+        _LOGGER.debug("Downloading file %s", filename)
+        try:
+            details: FilesHandlerDownloadDetails = await self.__call_files_api(
+                method="GET",
+                path=f"/download_details/{storage_type}/{filename}",
+            )
+        except TimeoutError as err:
+            raise FilesError(
+                "Timeout reached while trying to fetch download details",
+            ) from err
+        except ClientError as err:
+            raise FilesError(f"Failed to fetch download details: {err}") from err
+        except Exception as err:
+            raise FilesError(
+                f"Unexpected error while fetching download details: {err}",
+            ) from err
+
+        try:
+            response = await self._cloud.websession.get(
+                details["url"],
+                timeout=ClientTimeout(connect=10.0, total=_FILE_TRANSFER_TIMEOUT),
+            )
+
+            self._do_log_response(response)
+
+            response.raise_for_status()
+        except TimeoutError as err:
+            raise FilesError("Timeout reached while trying to download file") from err
+        except ClientError as err:
+            raise FilesError("Failed to download file") from err
```

Comment:
Maybe include the response status code in the error message? Ie catch `ClientResponseError` separately and then `ClientError`. For the `ClientError` case maybe include the exception in the message string?

Reply:
Not suited for MVP.

Reply:
Adding the exception to the message for `ClientError` is just a change on this line.

Reply:
It is not a change on this line, it is a new block, that also need to have tests for it as well.

Reply:
This is not a new block:

> For the ClientError case maybe include the exception in the message string?

Reply:
```suggestion
            raise FilesError(f"Failed to download file: {err}") from err
```

Reply:
That will leak sensitive data as aiohttp just spew out everything.

Reply:
Ok.